libsweep/src/sweep.cc

```diff
@@ -395,38 +262,132 @@ void sweep_device_set_sample_rate(sweep_device_s device, int32_t hz, sweep_error
   uint8_t args[2] = {0};
   sweep::protocol::integral_to_ascii_bytes(code, args);
 
-  sweep::protocol::error_s protocolerror = nullptr;
+  sweep::protocol::write_command_with_arguments(device->serial, sweep::protocol::SAMPLE_RATE_ADJUST, args);
 
-  sweep::protocol::write_command_with_arguments(device->serial, sweep::protocol::SAMPLE_RATE_ADJUST, args, &protocolerror);
+  sweep::protocol::response_param_s response;
+  sweep::protocol::read_response_param(device->serial, sweep::protocol::SAMPLE_RATE_ADJUST, &response);
+}
 
-  if (protocolerror) {
-    *error = sweep_error_construct("unable to send sample rate command");
-    sweep::protocol::error_destruct(protocolerror);
-    return;
-  }
+void sweep_device_reset(sweep_device_s device) {
+  SWEEP_ASSERT(device);
+  SWEEP_ASSERT(!device->is_scanning);
 
-  sweep::protocol::response_param_s response;
-  sweep::protocol::read_response_param(device->serial, sweep::protocol::SAMPLE_RATE_ADJUST, &response, &protocolerror);
+  sweep::protocol::write_command(device->serial, sweep::protocol::RESET_DEVICE);
+}
 
-  if (protocolerror) {
-    *error = sweep_error_construct("unable to receive sample rate command response");
-    sweep::protocol::error_destruct(protocolerror);
-    return;
+template <class F> auto translateException(F&& f, sweep_error_s* error) -> decltype(std::forward<F>(f)()) {
+  try {
+    return std::forward<F>(f)();
+  } catch (const sweep::ErrorBase& e) {
+    *error = new sweep_error{e.what()};
+  } catch (const std::bad_alloc&) {
+    *error = new sweep_error{"Could not allocate enough memory"};
+  } catch (...) {
+    *error = new sweep_error{"Unknown exception"};
+  }
+  return {};
+}
+
+template <class F> void translateExceptionV(F&& f, sweep_error_s* error) {
```

Comment:
As I said in the other thread: If you see a simple way to do this be my guest. also I really don't see what the advantage would be.